models/concerns/validate_data.rb

```diff
@@ -333,12 +322,28 @@ def check_relations(periodic_heuristic)
     check_relation_consistent_timewindows
     check_sticky_relation_consistency
 
-    incompatible_relation_types = @hash[:relations].collect{ |r| r[:type] }.uniq - %i[force_first never_first force_end same_vehicle]
-    return unless periodic_heuristic && incompatible_relation_types.any?
+    @hash[:relations]&.each{ |relation|
```

Comment:
I didn't quite understand. This is still the actual data sent by the user, we are at the json validation stage, so the raise corresponds exactly to the real data.

But more importantly, even if we are calling periodic heuristic, periodic heurictic cannot handle `Services in the same relation having different visits_number and/or min/max_lapse` (this was what you told me and this is why these checks are here). I remember having discussed with you that we cannot really expand the relations of such mismatching services..

Reply:
**Scheduling** can not handle those services, **periodic** can not handle relations at all. That is why what I was suggesting is that if we are going to call **periodic** then we raise with ```"#{incompatible_relation_types} relations not available with specified first_solution_strategy"``` message as soon as we detect that there are relations. If we are only using *scheduling*/*schedule* then we can do the checks you wrote.

Reply:
Now I understand ;) will move this in a `not periodic` condition

Reply:
related but separate question: what about the relations like `force_first never_first force_end same_vehicle` (i.e., the ones that periodic can manage. we want this check for these relations, right? 

I mean same_vehicle is obvious (since if multiple_services with different visits_number and/or max/min_lapse are in same_vehicle; we cannot expand)

What about the others? Can periodic expand such relations correctly (even if the services are "mismatching")?

Reply:
Those relations are particular because they used to be transformed into ```activity[:position]``` and in periodic heuristic we use non expanded services and apply constraints on all visits of this service. That is, if a service has ```position == :force_first``` then all of its visits will have this constraint.

In fact, those relations (force_first, never_first, force_end) do not need this check, because all services of the relation are considered individually.

Reply:
so we need to 
i. check same_vehicle even if it is periodic and
ii. skip checking some relations even in the scheduling case. is there any other such relations that doesn't need this check?

Reply:
In fact same_vehicle does not need check on visits either. Here is the list of all our existing relations, categorized according to check need : 

No check needed at all : 
same_vehicle 
same_route 
force_first 
never_first 
force_end
vehicle_trips
vehicle_group_duration vehicle_group_duration_on_weeks
vehicle_group_duration_on_months vehicle_group_number

Check needed if not periodic (reject if periodic) : 
sequence
order
shipment
meetup
minimum_day_lapse, maximum_day_lapse (these exist throught service.minimum and service.maximum lapse but if user provides minimum_day_lapse with a relation, periodic heuristic does not consider it for now),
minimum_duration_lapse, maximum_duration_lapse

Reply:
I was under the impression that `same_route` needs this check. If there are multiple services with mismatching lapse or visits_numbers in the same `same_route` relation, how do we manage it? 

This is why I mentioned `same_vehicle` because when there is no partitioning by vehicle, we treat `same_vehicle` as if it is `same_route` (if my memory serves me well, that was what you implemented, right?).

Reply:
Hum, in fact this is tricky because same_vehicle does not need the check because _same_vehicle_ implies that _original_vehicle_id_ should be the same but the day can be different. In this case we do not need to check visits.

But you are right regarding _same_route_, we need to do the check for _same_route_ relations. So we need to make sure that _same_vehicle_ is transformed into _same_route_ relation **before** this check. I think this is already the case.

Reply:
done and done :crossed_fingers: